Refactor LPUART driver constructors

JanKomarekNXP opened the pull request:
Updated constructors based on #22 

Changes:
- Add new_with_rtscts(), new_with_rts(), new_with_cts() methods
- Store RTS/CTS pins in TX/RX structs
- Remove enable_tx, enable_rx, enable_rx_rts, enable_tx_cts config fields
- Fix typo: msb_firs -> msb_first
- Fix write_byte() return type to Result<()>

JanKomarekNXP committed:
Refactor LPUART driver constructors

- Add new_with_rtscts(), new_with_rts(), new_with_cts() methods
- Store RTS/CTS pins in TX/RX structs
- Remove enable_tx, enable_rx, enable_rx_rts, enable_tx_cts config fields
- Fix typo: msb_firs -> msb_first
- Fix write_byte() return type to Result<()>

JanKomarekNXP committed:
Update hello example for LPUART driver changes

jamesmunns commented on src/lpuart/buffered.rs:
```diff
@@ -245,17 +180,139 @@ impl<'a> BufferedLpuart<'a> {
+        )?;
+
+        Ok(Self {
+            tx: tx.expect("TX should be created when tx_pin is provided"),
```
I don't feel strongly about it, but it's likely SOME day we'll come back to this to try to remove the potential panic here.

It MIGHT be worth thinking about splitting up `new_inner` in a way that allows you to independently configure the hardware (maybe `setup_core`, `setup_tx`, `setup_rx` or something), so we can rule out a panicking path at compile time.

That being said: this is an internal API, so it's not so big of a deal. We can kick the can down the road, or if you're interested, it might be worth cleaning up now.

JanKomarekNXP replied:
Totally agree that it does not look good. I have split the logic into a common init function and new_inner for Tx/Rx/Both. In my opinion, it looks nicer, and we got rid of the expect().

jamesmunns reviewed (approved):
Overall looks good to me, let me know if you have opinions about https://github.com/OpenDevicePartnership/embassy-mcxa/pull/51/files#r2569696866, but happy to smash the merge button whenever you are :)

jamesmunns commented on src/lpuart/buffered.rs:
```diff
@@ -58,156 +58,53 @@ pub struct BufferedLpuartTx<'a> {
-        unsafe { state.tx_buf.init(tx_buffer.as_mut_ptr(), tx_buffer.len()) }
+        // Initialize buffers
+        if let Some(tx_buffer) = tx_buffer {
+            assert!(!tx_buffer.is_empty());
```
We do have a Result here, we could return an err instead of asserting.

JanKomarekNXP replied:
Fixed in https://github.com/OpenDevicePartnership/embassy-mcxa/pull/51/commits/125494177186f664c20e5149b81e7adff5f33778

jamesmunns commented on src/lpuart/buffered.rs:
```diff
@@ -272,51 +329,58 @@ impl<'a> BufferedLpuart<'a> {
+            config,
+        )?;
+
+        tx.ok_or(Error::InvalidArgument)
```
Can this ever ACTUALLY be an Invalid argument? Or is this just the thing that you `expect` on elsewhere?

JanKomarekNXP replied:
Fixed in https://github.com/OpenDevicePartnership/embassy-mcxa/pull/51/commits/125494177186f664c20e5149b81e7adff5f33778

jamesmunns reviewed (commented):
A couple more nits about results :)

JanKomarekNXP committed:
Refactor buffered LPUART initialization logic

- Split new_inner() into init_common() and separate helpers for full-duplex, TX-only, and RX-only
- Replace assert!() with proper Error::InvalidArgument returns for empty buffers
- Simplify constructor implementations by removing expect() calls

jamesmunns commented:
Thanks @JanKomarekNXP!
